client/client.go

```diff
@@ -477,6 +482,11 @@ type (
 		// API. If the check fails, an error is returned.
 		CheckHealth(ctx context.Context, request *CheckHealthRequest) (*CheckHealthResponse, error)
 
+		// UpdateWorkflowExecution issues an update request to the specified
+		// workflow execution and returns the result synchronously.
+		// NOTE: Experimental
+		UpdateWorkflowExecution(ctx context.Context, workflowID string, updateName string, args []interface{}, opts UpdateWorkflowExecutionOptions) (converter.EncodedValue, error)
```

Comment:
Signal and query both accept can-be-empty `runID` option in the signature. I wonder if we should too.

Now I'm wondering, with first execution run ID being the only option that should appear in that struct (runID can be in signature and headers are interceptor only), and since cancel/terminate _also_ offer first execution run ID in their proto request and yet never allow a user to set it, is there any value for options here at all? We can then use varargs (so nice) and we will match signal, query, cancel, terminate, etc.

I think if we want to come back one day and add first execution run ID, we'd do it to multiple calls, not just this one.

So, despite our Slack convo and against known future proofing, I'm not on "team no options". Thoughts?

Reply:
Yeah, I was going go make a note of that. I don't like the difference but I also don't like optional fields so both have downsides and I'm not strongly attached to either approach. 

If we don't allow FirstExecutionRunID, what are the chances we regret it in the future? I'm not familiar enough with how often it's used in practice. I guess if we end with ...interface{} then we could also allow for well-known option types for extensibility in the future. Gross but at least it's possible.

Reply:
> If we don't allow FirstExecutionRunID, what are the chances we regret it in the future?

We haven't regretted it yet on cancel or terminate

> I guess if we end with ...interface{} then we could also allow for well-known option types for extensibility in the future. Gross but at least it's possible.

Gross indeed. Another option of course is an overload, e.g. `UpdateWorkflowExecutionWithOptions`. This is what we had to do with `QueryWorkflow`. Note there is a `QueryWorkflowWithOptions` and it _only_ takes options and has everything inside of there. I'd support that. So maybe:

```go
UpdateWorkflow(ctx context.Context, workflowID string, runID string, updateName string, args ...interface{}) (converter.EncodedValue, error)

UpdateWorkflowWithOptions(ctx context.Context, req *UpdateWorkflowWithOptionsRequest) (*UpdateWorkflowWithOptionsResponse, error)
```

I think this, while admittedly ugly, is the best experience for users at the cost of just one overload.

Note I also changed `UpdateWorkflowExecution` to just `UpdateWorkflow` to match start, cancel, terminate, signal, and query (and deviating from describe and reset by intention, those are bad APIs that should never be exposed w/ that raw proto mess).

Reply:
I'm fine with the two function approach. I wish we had don functional args way back when.

FYI the gRPC API definition will be keeping the "Execution" part. But we already have prior art for diverging from the service's naming patterns with workflowservice.SignalWorkflowExecution vs. client.SignalWorkflow.

Reply:
Definitely want "Execution" on the gRPC request call itself, that very much matches the others.